### PR TITLE
feat: Certificate Issuing flow (part 3) WPB-3290

### PR DESCRIPTION
**Pull request by KaterinaWire**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-3290" title="WPB-3290" target="_blank">StoryWPB-3290</a>  [iOS] Implement Certificate Issuing flow [except for challenges]
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

This part contains many steps for certificate enrolling a certificate, including challenges validation. Some of these steps required requests to the Wire server, and I used ApiProvider to access them.
Unfortunately I can't use existing HttpClient to make acme requests and I need to keep `HttpClientE2EI` for now.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-request-strategy/Sources/E2EIdentity/ACME/AcmeAPITests.swift [hunk 1]
```diff
@@ -224,6 +222,38 @@ class AcmeClientTests: ZMTBaseTest {
...
+        let path = "https://acme.example.com/acme/provisioner1/challenge/foVMOvMcapXlWSrHqu4BrD1RFORZOGrC/1pceubrFUZAvVQI5XgtLDMfLefhOt4YI"
+
+        // mock
+        let mockResponse = HTTPURLResponse(
+            url: URL(string: path)!,
+            statusCode: 200,
+            httpVersion: "",
+            headerFields: ["Replay-Nonce": headerNonce]
+        )!
+        let challengeResponseData = try! JSONEncoder.defaultEncoder.encode(expectation)
```
Comment (David-Henner) [inline wire-ios-request-strategy/Sources/E2EIdentity/ACME/AcmeAPITests.swift]: remove force unwrapping to ensure tests don't crash
```suggestion
        let mockResponse = try XCTUnwrap(HTTPURLResponse(
            url: URL(string: path)!,
            statusCode: 200,
            httpVersion: "",
            headerFields: ["Replay-Nonce": headerNonce]
        ))
        let challengeResponseData = try JSONEncoder.defaultEncoder.encode(expectation)
```

File: wire-ios-request-strategy/Sources/E2EIdentity/EnrollE2eICertificateUseCase.swift [hunk 2]
```diff
@@ -42,9 +49,44 @@ public final class EnrollE2eICertificateUseCase: EnrollE2eICertificateUseCaseInt
...
+                                                                               prevNonce: dpopChallengeResponse.nonce,
+                                                                               acmeChallenge: oidcChallenge)
+
+        let orderResponse = try await enrollment.checkOrderRequest(location: newOrder.location, prevNonce: oidcChallengeResponse.nonce)
+        let finalizeResponse = try await enrollment.finalize(location: orderResponse.location, prevNonce: orderResponse.acmeResponse.nonce)
+        let certificateRequest = try await enrollment.certificateRequest(location: finalizeResponse.location, prevNonce: finalizeResponse.acmeResponse.nonce)
+
+        do {
+            return try JSONDecoder().decode(String.self, from: certificateRequest.response)
+            // certificateRequest.response.payloadString()
```
Comment (typfel) [inline wire-ios-request-strategy/Sources/E2EIdentity/EnrollE2eICertificateUseCase.swift]: Delete?

File: wire-ios-request-strategy/Sources/APIs/E2eIAPI.swift [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                                         payload: nil,
+                                         apiVersion: apiVersion.rawValue)
+        request.addValue(dpopToken, forAdditionalHeaderField: Constant.dpopHeaderKey)
+        let response = await httpClient.send(request)
+
+        guard let accessToken = AccessTokenResponse(response) else {
+            throw NetworkError.errorDecodingResponse(response)
+        }
+        return accessToken
+    }
```
Comment (typfel) [inline wire-ios-request-strategy/Sources/APIs/E2eIAPI.swift]: Can you use this helper function?

```suggestion
        mapResponse(response)
    }
```

**Review by typfel — CHANGES_REQUESTED** [review]

**Review by David-Henner — APPROVED** [review]

File: wire-ios-request-strategy/Sources/E2EIdentity/EnrollE2eICertificateUseCase.swift (hunk 2 above)
Comment (KaterinaWire) [reply wire-ios-request-strategy/Sources/E2EIdentity/EnrollE2eICertificateUseCase.swift]: deleted

**Review by typfel — APPROVED** [review]

**Pull request merged**